### ISSUE TITLE
ProxyTransactionManagementConfiguration not eligible for auto-proxying

### ISSUE THREAD

**Issue by Marco-Sulla** [spring-projects/spring-boot#17561]
Using `Spring Boot` 2.1.6.RELEASE, I get this warning

```log
Bean 'org.springframework.transaction.annotation.ProxyTransactionManagementConfiguration' of type [org.springframework.transaction.annotation.ProxyTransactionManagementConfiguration$$EnhancerBySpringCGLIB$$efb8db88] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
```
I checked `ProxyTransactionManagementConfiguration` in your code and it does not contains any `@Autowired`. But the strange fact is that if I select the branch `v2.1.6.RELEASE`, no file `ProxyTransactionManagementConfiguration` exists.

**Comment by wilkinsona** [comment]
`ProxyTransactionManagementConfiguration` is part of Spring Framework. You can find its source code [here](https://github.com/spring-projects/spring-framework/blob/e92cbe193807599624e01cfd37807cbba9d8cd12/spring-tx/src/main/java/org/springframework/transaction/annotation/ProxyTransactionManagementConfiguration.java).

The error that you are seeing indicates that something has caused `ProxyTransactionManagementConfiguration` to be initialised too early. Specifically, it has been initialised before all bean post-processors have been created. This prevents it from being post-processed.

If you would like us to spend some time investigating why that's happening, we'll need some more information. Can you please provide a minimal sample (either as a zip attached to this issue or in a separate repository) that reproduces the behaviour that you have described?

**Comment by Marco-Sulla** [comment]
Yes, excuse me, I wrong the project, and furthermore it seems it's fixed now in trunk:

`@Configuration(proxyBeanMethods = false)`

**Comment by wilkinsona** [comment]
> `@Configuration(proxyBeanMethods = false)`

Just for the record, that setting won't have any effect on an early initialization problem preventing post-processing of the `ProxyTransactionManagementConfiguration` bean.

**Comment by ybulatnikov** [comment]
Creating project from scratch with these dependencies:

![image](https://user-images.githubusercontent.com/48504701/62227499-b4375d80-b3c4-11e9-9a97-8720f84e1c52.png)

Running maven test (without any changes in code):

![image](https://user-images.githubusercontent.com/48504701/62227874-5d7e5380-b3c5-11e9-931a-30bff42a0a37.png)

**Comment by EricLiu** [comment]
This problem can be repeated in a sample spring-boot project create by Initializr.
pom.xml like this:

<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.2.0.M5</version>
		<relativePath/> 
	</parent>
	<groupId>debug</groupId>
	<artifactId>demo</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<name>demo</name>
	<description>Demo project for Spring Boot</description>

	<properties>
		<java.version>1.8</java.version>
	</properties>

	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-data-jpa</artifactId>
		</dependency>
		<dependency>
			<groupId>org.postgresql</groupId>
			<artifactId>postgresql</artifactId>
			<version>42.2.6</version>
		</dependency>
	</dependencies>

	<build>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		</plugins>
	</build>
	<repositories>
		<repository>
			<id>spring-milestones</id>
			<name>Spring Milestones</name>
			<url>https://repo.spring.io/milestone</url>
		</repository>
	</repositories>
	<pluginRepositories>
		<pluginRepository>
			<id>spring-milestones</id>
			<name>Spring Milestones</name>
			<url>https://repo.spring.io/milestone</url>
		</pluginRepository>
	</pluginRepositories>
</project>

And application class like this:
----------------------------
package debug.demo;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Configuration;

@SpringBootApplication()
public class DemoApplication {

	public static void main(String[] args) {
		SpringApplication.run(DemoApplication.class, args);
	}

}

**Comment by snicoll** [comment]
@EricLiu I've played around with your instructions and noticed this message doesn't show up if Spring Data is removed and replaced by the core JDBC starter (that still configures a transaction manager). I suspect Spring Data's initialization might trigger the early init and I've asked @mp911de to look into it.

**Comment by ybulatnikov** [comment]
The easiest way to reproduce, I think, is to generate project via start.spring.io with 3 dependencies: 

spring-boot-starter-web 
spring-boot-starter-data-jpa 
h2 (as runtime database)

then without any changes (or any additional configuration) run the project or run test.

Also I've noticed this is not reproducible if I change starter-web to starter-webflux.

**Comment by wilkinsona** [comment]
Thank you, @ybulatnikov. That's consistent with @EricLiu's comment above. As @snicoll mentioned, @mp911de from the Spring Data team is looking into it.

**Comment by naspredam** [comment]
I can see that this is fixed on latest version of 2.2.x.

There is any tentative to fix on Using Spring Boot 2.1.x? @snicoll
